0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+
+Of course, it expands the language, where a code solution is sufficient, if not elegant.
+
+Additionally,
+this would be the first `macro_rules!` macro where a capture isn't plain token (tree) substitution.
+If the `match` trick is actually fully equivalent to function argument semantics,
+(the author believes this to be true but hasn't exhaustively shown this,)
+then tools such as `cargo expand` (`-Z unpretty`) can use it as the macro expansion.
+But still, this moves `macro_rules!` away from just being token (tree) substitution,
+and gives `macro_rules!` a superpower that isn't available to function-like proc macros.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Of course,
+we could just not do this,
+and `macro_rules!` authors can just use the `match` trick.
+More importantly, though,
+there are two other discussed features that could address the same problem space:
+
+## `$:place`
+[macro_place]: #macro-place
+
+For exposition, we use
+`macro_rules! m { ( $x:value ) => ( &$x ); }`
+to explain functionality.
+
+`$:value` as described above generates a new _named_ temporary for the captured value,
+to match the behavior of function arguments exactly.
+That is,
+`m!(array[0])` would call `Deref::deref` and return a reference to a copy of the deref'd to value
+(which would be dropped immediately, causing a borrowck error if bound)
+exactly matching the behavior of a function, which would fail typeck.
+
+An alternative semantic, which I call `$:place`,
+captures the _place_ in this situation, not the _value_.
+That is, `m!(array[0])` would call `Deref::deref` and return that reference directly.
+If `m!` were defined as `=> ( &mut $x );`,
+then `Deref::deref_mut` would be called.
+If the capture is expanded only once,
+this behaves identically to an `$:expr` capture,
+except for the evaluation timing of side effects.
+
+However, `$:place` adds a new concept to Rust,
+that of capturing a _place_ directly.
+This is entirely impossible in surface Rust today.
+This form of capturing may be more intuitive to macro authors
+(who are already used to and use similar behavior from `$:expr`).
+However,
+as this is much more complicated on the implementation side than a simple `$:value`,
+and `$:value` offers most of the benefit of `$:place` without the extra implementation complexity,
+`$:place` is just offered here as an alternative.
+
+## `macro fn`
+[macro_fn]: #macro-fn
+
+Another possibility that's been discussed is `macro fn`.
+Basically, these would be `fn`, and have the semantics of `fn`,
+but be duck typed (like macros) and semantically copy/pasted into the calling scope.
+This is basically the exact feature that "`macro_rules!` with function-like captures" is trying to serve,
+except for one important thing: a `macro fn` is (potentially) still a `fn` in that it has one fixed airity,
+and can't be overloaded like a macro can.
+Basically, `macro fn` is asking for "macros 2.0,"
```

Comment:
Spurious comma.

Reply:
This comma is part of

> asking for macros 2.0, which is still desirable, but still 

so isn't it required here? (Does British English put the comma outside the quotes? I know the American English rules put the comma inside the quotes (and the Rust project generally prefers American English).)

Reply:
@CAD97 In technical writing, it's entirely normal to always put punctuation outside the quotes if it's not part of what's being quoted, following the British English quoting style even in otherwise American English writing.